docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,247 @@
+
+```golang
+// GetDenom retreives the full identifiers trace from the store.
+func (k Keeper) GetTrace(ctx Context, traceHash []byte) string {
+  store := ctx.KVStore(k.storeKey)
+  bz := store.Get(types.KeyTrace(traceHash))
+  if len(bz) == 0 {
+    return ""
+  }
+  return string(bz)
+}
+
+// HasTrace checks if a the key with the given trace hash exists on the store.
+func (k Keeper) HasTrace(ctx Context, traceHash []byte)  bool {
+  store := ctx.KVStore(k.storeKey)
+  return store.Has(types.KeyTrace(traceHash))
+}
+
+// SetTrace sets a new {trace hash -> trace} pair to the store.
+func (k Keeper) SetTrace(ctx Context, traceHash []byte, trace string) {
+  store := ctx.KVStore(k.storeKey)
+  store.Set(types.KeyTrace(traceHash), []byte(trace))
+}
+```
+
+When a fungible token is send to a sink chain, the trace information needs to be updated with the new port and channel identifiers:
+
+```golang
+// PrefixDenom adds the given port and channel identifiers prefix to the denomination and sets the
+// new {trace hash -> trace} pair to the store.
+func (k Keeper) PrefixDenom(ctx Context, portID, channelID, denom string) string {
+  // Get each component of the denom. The resulting slice will be:
+  //
+  // - [ "ibc", traceHash, baseDenom], if the denom is dirty (contains trace metadata).
+  // - [ baseDenom ], if the denom has never been sent from the origin chain.
+  denomSplit := strings.Split(denom, "/")
+
+  var (
+    baseDenom string
+    trace     string
+    traceHash tmbytes.HexBytes
+  )
+
+  // check if the denomination is clean or if it contains the trace info
+  if denomSplit[0] == denom {
+    baseDenom = denom
+    trace = portID + "/" + channelID +"/"
+  } else {
+    // concatenate if base denom already contains a separator
+    baseDenom = strings.Join(denomSplit[2], "/")
+    traceHash = tmbytes.HexBytes(denomSplit[1])
+    // Get the value from the map trace hash -> denom identifiers prefix
+    trace = k.GetTrace(ctx, traceHash)
+    // prefix the identifiers to create the new trace
+    trace = portID + "/" + channelID +"/" + trace + "/"
+  }
+  
+  traceHash = tmbytes.HexBytes(tmhash.Sum(trace))
+
+  // set the value to the lookup table if not stored already
+  if !k.HasTrace(ctx, traceHash) {
+    k.SetTrace(ctx, traceHash)
+  }
+
+  denom = "ibc/"+ traceHash.String()  "/" + baseDenom
+  return denom
+}
+```
+
+The denomination also needs to be updated when token is received on the source chain:
+
+```golang
+// UnprefixDenom removes the first portID/channelID pair from a given denomination trace info and returns the
+// denomination with the updated trace hash and the new trace info.
+// An error is returned if the trace cannot be found on the store from the denom's trace hash.
+func (k Keeper) UnprefixDenom(ctx Context, denom string) (denom, trace string, err error) {
+  denomSplit := strings.Split(denom, "/")
```

Comment:
`SplitN`? and then remove Join?

Reply:
👍 I forgot about that function was available too